File: terminal/viewport.py
from viewer import ViewerConstants
from state import State
from interactive import VisibleModule, FeatureModule, SavableModule, InteractiveModule
from status_line import StatusLine
import colors
import features
import curses
import log
import math

log = log.logger

# TODO: Implement save method

class Viewport(VisibleModule, FeatureModule, SavableModule, InteractiveModule):
    def __init__(self):


        self.initial_draw_priority = 0
        self.draw_priority = 1

        self.x = 0
        self.y = 0
        self.h = 100
        self.w = 100

        self._features = []
        self._units = []
        self._fow = [ [ False for y in range(self.h) ] for x in range(self.w) ]

        self.cursor_y = math.floor(self.h/4)
        self.cursor_x = math.floor(self.w/2)

        self.box_xy = None

        self._screen = curses.newpad(self.h, self.w)
        self._dirty = True

    def draw(self, viewer, force=False):
        if self._dirty or force:
            if force: log.debug("viewport.draw forced")
            if self._dirty: log.debug("viewport is dirty")
            self._screen.erase()

            self._screen.attrset(colors.get("White"))
            self._screen.border(
                    curses.ACS_BOARD,
                    curses.ACS_BOARD,
                    curses.ACS_BOARD,
                    curses.ACS_BOARD,
                    curses.ACS_BOARD,
                    curses.ACS_BOARD,
                    curses.ACS_BOARD,
                    curses.ACS_BOARD
            )
            self._screen.attroff(colors.get("White"))

            [ features.draw(viewer, self._screen, feature, self.x, self.y, self.h, self.w) for feature in self._features ]

            [ unit.draw(viewer, self._screen) for unit in self._units ]


            state = viewer.get_submodule(State)
            if state.get_state("role") == "pc" or (state.get_state("role") == "gm" and state.get_state("fow") == "on"):
                for x in range(0, self.w-2):
                    for y in range(0, self.h-2):
                        if self._fow[x][y]:
                            self._screen.addstr(y+1, x+1, "▒", colors.get("Dark Grey"))

            if self.box_xy != None:
                x_min = min(self.box_xy[0], self.cursor_x)
                x_max = max(self.box_xy[0], self.cursor_x) + 1

                y_min = min(self.box_xy[1], self.cursor_y)
                y_max = max(self.box_xy[1], self.cursor_y) + 1

                for y in range(y_min, y_max):
                    for x in range(x_min, x_max):
                        self._screen.addstr(y, x, "▒", colors.get("Orange"))

            self._screen.addch(self.cursor_y, self.cursor_x, ord('X'), colors.get("Gold"))

            self._screen.noutrefresh(
                    self.y,
                    self.x,
                    0, math.floor(ViewerConstants.max_x/3),
                    ViewerConstants.max_y-3,
                    math.floor(ViewerConstants.max_x/2)+math.floor(ViewerConstants.max_x/3))
            self._dirty = False
            return True
        return False

    def _handle(self, viewer, ch):

        if ch == ord("+") or ch == ord("="):

            unit = self.get_current_unit()
            if unit:
                unit.current_health += 1

                self.client.send({
                    "type": "command",
                    "key": "modify.map.unit",
                    "details": unit.toDict()
                })


        elif ch == ord("-"):
            unit = self.get_current_unit()
            if unit:
                unit.current_health -= 1

                self.client.send({
                    "type": "command",
                    "key": "modify.map.unit",
                    "details": unit.toDict()
                })



    def _handle_help(self, viewer, buf):
        pass

    def _handle_combo(self, viewer, buf):
        pass

    def right(self):
        if self.w - self.x > math.floor(ViewerConstants.max_x/2)+1:
            self.x += 1
            self.cursor_right()
            self._dirty = True

    def down(self):
        if (self.h - self.y) > ViewerConstants.max_y-2:
            self.y += 1
            self.cursor_down()
            self._dirty = True

    def up(self):
        if self.y-1 >= 0:
            self.y -= 1
            self.cursor_up()
            self._dirty = True

    def left(self):
        if self.x-1 >= 0:
            self.x -= 1
            self.cursor_left()
            self._dirty = True

    def cursor_up(self):
        if self.cursor_y-1 >= 1:
            self.cursor_y -= 1
            self._dirty = True

    def cursor_down(self):
        if ( self.cursor_y + 1 < ViewerConstants.max_y-2 + self.y and
             self.cursor_y + 1 < self.h-1 ):
            self.cursor_y +=1
            self._dirty = True

    def cursor_right(self):
        if ( self.cursor_x + 1 < math.floor(ViewerConstants.max_x/2) +1 + self.x and
             self.cursor_x + 1 < self.w-1 ):
            self.cursor_x +=1
            self._dirty = True

    def cursor_left(self):
        if self.cursor_x-1 >= 1:
            self.cursor_x -= 1
            self._dirty = True

    def get_cursor_focus(self, viewer):
        state = viewer.get_submodule(State)
        unit = [ unit for unit in self._units if unit.y == self.cursor_y and unit.x == self.cursor_x ]
        feature = [feature for feature in self._features if feature["y"] == self.cursor_y and feature["x"] == self.cursor_x]

        if not (state.get_state("role") == "pc" and self._fow[self.cursor_x][self.cursor_y]):
        # ensures there is a feature and if we are a pc we will not show thing hidden by FoW
            if len(unit):
                sl = viewer.get_submodule(StatusLine)
                                #name + space         border []
                text = "%s %s/%s " % (unit[0].name, unit[0].current_health, unit[0].max_health)
                bar = ""

                if unit[0].max_health != 0:
                    width = sl.w - (len(text) + 1) - 2 - 2

                    percent = unit[0].current_health/float(unit[0].max_health)

                    if percent <= 1.0 and percent >= 0.0:
                        number_of_units = math.ceil(percent * width)
                        bar = number_of_units * "="
                        bar = bar.rjust(width, " ")

                    elif percent > 1.0:
                        diff = percent - 1.0
                        diff_percent = diff/percent
                        excess = math.ceil(width * diff_percent)

                        bar = excess * "="
                        bar += "(+)|"
                        bar = bar.ljust(width, "=")

                    elif percent < 0.0:
                        diff = (-1 * percent)/(1.0 + ( -1 * percent))
                        log.error("diff %s" % diff)
                        excess = math.ceil(width * diff)

                        bar = "|(-)"
                        bar += excess * "="
                        bar = bar.rjust(width, " ")


                desc = "%s [%s]" % (text, bar)
                return desc

            elif(len(feature)):
                desc = "%s" % feature[0]["type"]
                return desc

        return ""


    def get_feature_idx(self, y, x):
        for feature in self._features:
            if feature["y"] == y and feature["x"] == x:
                return self._features.index(feature)

    def get_feature(self, idx):
        return self._features[idx]

    def update_feature(self, idx, feature):
        self._features[idx] = feature

    def serialize_features(self):
        features = []
        for feature in self._features:
            features.append(feature)
        return features

    def update_features(self, feature_dicts):
        self._features = []
        for feature_dict in feature_dicts:
            self._features.append(feature_dict)

        if len(feature_dicts):
            self._dirty = True

    def set_fow(self, new_fow):
        self._fow = new_fow
        self._dirty = True

    def add_fow(self, point):
        self._fow[point["x"]][point["y"]] = True
        self._dirty = True

    def remove_fow(self, point):
        log.info(("point", point))
        self._fow[point["x"]][point["y"]] = False
        self._dirty = True

    def update_units(self, units):
        self._units = units
        self._dirty = True

    def update_screen(self, max_y, max_x):
        self.h = max_y + 1
        self.w = max_x + 2

        del self._screen
        self._screen = curses.newpad(self.h, self.w)

    def get_current_unit(self):
        for unit in self._units:
            if unit.y == self.cursor_y and unit.x == self.cursor_x:
                return unit
        return None

File: web/viewport.js
"use strict";

var viewport = {};

viewport.init = function(){

    cursed.modules.interactive.push(viewport);
    
    viewport.x = 0 
    viewport.y = 0;
    viewport.width = cursed.grid.width;
    viewport.height = cursed.grid.height;

    viewport.cursor_x = 10;
    viewport.cursor_y = 10;

    // virtual width and height. should be set to map h and w
    viewport.v_x = 0;
    viewport.v_y = 0;
    viewport.v_height = viewport.height;
    viewport.v_width = viewport.width; 

    viewport.features = [];
    viewport.fow = [];
    viewport.units = [];
    viewport.box_xy = null;

    viewport.dirty = true;

    var diff = new Array(cursed.grid._.length);
    for(var i=0; i<cursed.grid._.length; i++){
        diff[i] = new Array(cursed.grid._[0].length);
        for(var j=0; j<cursed.grid._[0].length; j++){
            diff[i][j] = null;
        }
    }
    viewport.last = diff

    viewport.clear();
}

viewport.draw = function(){

    if(viewport.dirty){

        var diff = new Array(cursed.grid._.length);
        for(var i=0; i<cursed.grid._.length; i++){
            diff[i] = new Array(cursed.grid._[0].length);
            for(var j=0; j<cursed.grid._[0].length; j++){
                diff[i][j] = null;
            }
        }

        var keys = Object.keys(viewport.features);
        var len = keys.length;
        var i = 0;

        // Draw features
        while(i < len){
            var feature = viewport.features[keys[i]];
            if(feature.x >= viewport.v_x &&  /* feature is within visible bounds */
               feature.y >= viewport.v_y &&
               feature.x < (viewport.v_x + viewport.width) &&
               feature.y < (viewport.v_y + viewport.height) &&
               ( cursed.state.fow == "off" ||  /* fog of war is not off and not occluding feature */
                   ( viewport.fow[feature.x] !== undefined && !viewport.fow[feature.x][feature.y] )
               )){

                var raw_feature = cursed.features.get(feature.type);
                if(raw_feature !== null && raw_feature !== undefined){
                    var character = raw_feature.character;
                    if(character !== null && character !== undefined){
                        diff[feature.y - viewport.v_y][feature.x - viewport.v_x] = feature.type;
                        cursed.grid.text(feature.y - viewport.v_y, feature.x - viewport.v_x, character, raw_feature.color);
                    }
                }
            }

            i++;
        }

        // Draw units
        for(var i=0; i<viewport.units.length; i++){
            var unit = viewport.units[i];

            if(unit.x >= viewport.v_x && /* unit is within visible bounds */
               unit.y >= viewport.v_y &&
               unit.x < (viewport.v_x + viewport.width) &&
               unit.y < (viewport.v_y + viewport.height) &&
               ( cursed.state.fow == "off" ||  /* fog of war is not off and not occluding unit */
                   ( viewport.fow[unit.x] !== undefined && !viewport.fow[unit.x][unit.y] )
                )){

                var corrected_y = unit.y - viewport.v_y;
                var corrected_x = unit.x - viewport.v_x;
                diff[corrected_y][corrected_x] = "unit:" + unit.id;
                if(unit.controller == cursed.state.username  && cursed.state.role != "gm"){
                    cursed.grid.text(corrected_y, corrected_x, "@", "Light Blue");
                }
                else if(unit.type == "pc"){
                    cursed.grid.text(corrected_y, corrected_x, "@", "Light Green");
                }
                else if(unit.type == "enemy"){
                    cursed.grid.text(corrected_y, corrected_x, "@", "Light Red");
                }
                else if(unit.type == "neutral" || unit.type === ""){
                    cursed.grid.text(corrected_y, corrected_x, "@", "Light Grey");
                }
            }
        }


        // Draw fog of war
        if(cursed.state.fow == "on"){
            for(var x=0; x<viewport.fow.length; x++){
                for(var y=0; y<viewport.fow[x].length; y++){
                    if(viewport.fow[x][y] &&
                       x >= viewport.v_x && 
                       y >= viewport.v_y &&
                       x < (viewport.v_x + viewport.width) &&
                       y < (viewport.v_y + viewport.height)){

                        diff[y - viewport.v_y][x - viewport.v_x] = "FOW";
                        cursed.grid.text(y - viewport.v_y, x - viewport.v_x, "\u2588", "Blue Grey");
                    }
                }
            }
        }

        //clear blocks not in new rendering
        for(var i=0; i<diff.length; i++){
            for(var j=0; j<diff[i].length; j++){
                if( viewport.last[i][j] === diff[i][j] ){
                    cursed.grid.text(viewport.y + i, viewport.x + j, " ", "Gold");
                }
            }
        }

        //draw notes
        for(var note of cursed.notes.notes){
            if(note.x >= viewport.v_x && 
               note.y >= viewport.v_y &&
               note.x < (viewport.v_x + viewport.width) &&
               note.y < (viewport.v_y + viewport.height)){

                cursed.grid.text(note.y - viewport.v_y, note.x - viewport.v_x, "!", "Red");
            }
        }

        //draw select box if an xy are set
        if(viewport.box_xy !== null){

            var x_min = Math.min(viewport.box_xy[0], viewport.cursor_x);
            var x_max = Math.max(viewport.box_xy[0], viewport.cursor_x) + 1;
            
            var y_min = Math.min(viewport.box_xy[1], viewport.cursor_y);
            var y_max = Math.max(viewport.box_xy[1], viewport.cursor_y) + 1;

            for(var y=y_min; y<y_max; y++){
                for(var x=x_min; x<x_max; x++){
                    cursed.grid.text(viewport.y + y, viewport.x + x, "\u2588", "Orange");
                }
            }
        }

        // Draw Cursor
        cursed.grid.text(viewport.y + viewport.cursor_y, viewport.x + viewport.cursor_x, "X", "Gold");

        viewport.dirty = false;
        cursed.viewer.dirty = true;
    }
    
};

viewport.clear = function(){
    for(var y=0; y<viewport.height; y++){
        for(var x=0; x<viewport.width; x++){
            //cursed.grid.text(viewport.y + y, viewport.x + x, " ", "Gold");
        }
    }
}

viewport.handle_help = function(buff){

}

viewport.handle = function(event){
    if(!cursed.viewer.handling && 
        (cursed.command_window.mode === cursed.command_window.command_modes.default ||
         cursed.command_window.mode === cursed.command_window.command_modes.build ||
         cursed.command_window.mode === cursed.command_window.command_modes.fow ||
         cursed.command_window.mode === cursed.command_window.command_modes.initiative_unit_select ||
         cursed.command_window.mode === cursed.command_window.command_modes.box_select) &&
        !(cursed.command_window.box && cursed.command_window.mode != cursed.command_window.command_modes.box_select)  // prevents user from moving around when box is still showing but box screen is exited
    ){
    
        if(cursed.state.move_mode === "hjkl"){
            if(event.key === "j"){ viewport.cursor_down(); }
            else if(event.key === "k"){ viewport.cursor_up(); }
            else if(event.key === "h"){ viewport.cursor_left(); }
            else if(event.key === "l"){ viewport.cursor_right(); }

            if(cursed.command_window.mode != cursed.command_window.command_modes.box_select){
                if(event.key === "J"){ viewport.down(); }
                else if(event.key === "K"){ viewport.up(); }
                else if(event.key === "H"){ viewport.left(); }
                else if(event.key === "L"){ viewport.right(); }
            }
        }
        else if(cursed.state.move_mode === "ijkl"){
            if(event.key === "k"){ viewport.cursor_down(); }
            else if(event.key === "i"){ viewport.cursor_up(); }
            else if(event.key === "j"){ viewport.cursor_left(); }
            else if(event.key === "l"){ viewport.cursor_right(); }

            if(cursed.command_window.mode != cursed.command_window.command_modes.box_select){
                if(event.key === "K"){ viewport.down(); }
                else if(event.key === "I"){ viewport.up(); }
                else if(event.key === "J"){ viewport.left(); }
                else if(event.key === "L"){ viewport.right(); }
            }
        }
    }
        
    if(event.key == "f" && 
        cursed.state.role === "gm" &&
        (cursed.command_window.mode === cursed.command_window.command_modes.default ||
         cursed.command_window.mode === cursed.command_window.command_modes.fow) ){
        if(cursed.state.fow === "on"){
            cursed.state.fow = "off";
            cursed.viewport.dirty = true;
            cursed.viewport.clear();
            cursed.viewport.draw();
        }
        else if(cursed.state.fow == "off"){
            cursed.state.fow = "on";
            cursed.viewport.dirty = true;
            cursed.viewport.clear();
            cursed.viewport.draw();
        }
    }
    else if (event.key === "+" || event.key === "="){
        var unit = viewport.getCurrentUnit();
        
        if(unit !== null){
            var modified_unit = {
                name: unit.name,
                id: unit.id,
                current_health: unit.current_health + 1,
                max_health: unit.max_health,
                x: unit.x,
                y: unit.y,
                controller: unit.controller,
                type: unit.type
            }

            cursed.client.send({
                type: "command",
                key: "modify.map.unit",
                details: modified_unit
            }, true);
        }
    }
    else if (event.key === "-"){
        var unit = viewport.getCurrentUnit();
        
        if(unit !== null){
            var modified_unit = {
                name: unit.name,
                id: unit.id,
                current_health: unit.current_health - 1,
                max_health: unit.max_health,
                x: unit.x,
                y: unit.y,
                controller: unit.controller,
                type: unit.type
            }

            cursed.client.send({
                type: "command",
                key: "modify.map.unit",
                details: modified_unit
            }, true);
        }
    }
}

viewport.handle_combo = function(event){};

viewport.cursor_up = function(){
    cursed.viewer.handling = true;
    setTimeout(()=>{cursed.viewer.handling = false;}, 100);

    if(viewport.cursor_y > 0){
        cursed.grid.text(viewport.y + viewport.cursor_y, viewport.x + viewport.cursor_x, " ", "Gold");
        viewport.cursor_y -= 1;
        viewport.dirty = true;

        viewport.draw();

        cursed.status_line.dirty = true;
        cursed.status_line.draw();

        //cursed.command_window.dirty = true;
        //cursed.command_window.draw();
    }
}

viewport.cursor_down = function(){
    cursed.viewer.handling = true;
    setTimeout(()=>{cursed.viewer.handling = false;}, 100);

    if(viewport.cursor_y < viewport.height-1){
        cursed.grid.text(viewport.y + viewport.cursor_y, viewport.x + viewport.cursor_x, " ", "Gold");
        viewport.cursor_y += 1;
        viewport.dirty = true;

        viewport.draw();

        cursed.status_line.dirty = true;
        cursed.status_line.draw();

        //cursed.command_window.dirty = true;
        //cursed.command_window.draw();

    }
}

viewport.cursor_left = function(){
    cursed.viewer.handling = true;
    setTimeout(()=>{cursed.viewer.handling = false;}, 100);

    if(viewport.cursor_x > 0){
        cursed.grid.text(viewport.y + viewport.cursor_y, viewport.x + viewport.cursor_x, " ", "Gold");
        viewport.cursor_x -= 1;
        viewport.dirty = true;

        viewport.draw();

        cursed.status_line.dirty = true;
        cursed.status_line.draw();

        //cursed.command_window.dirty = true;
        //cursed.command_window.draw();
    }
}

viewport.cursor_right = function(){
    cursed.viewer.handling = true;
    setTimeout(()=>{cursed.viewer.handling = false;}, 100);

    if(viewport.cursor_x < viewport.width-1){
        cursed.grid.text(viewport.y + viewport.cursor_y, viewport.x + viewport.cursor_x, " ", "Gold");
        viewport.cursor_x += 1;
        viewport.dirty = true;

        viewport.draw();

        cursed.status_line.dirty = true;
        cursed.status_line.draw();

        //cursed.command_window.dirty = true;
        //cursed.command_window.draw();
    }
}

viewport.up = function(){
    cursed.viewer.handling = true;
    setTimeout(()=>{cursed.viewer.handling = false;}, 100);

    if(viewport.v_y > 2){
        viewport.v_y -= 2;
        viewport.dirty = true;
        viewport.clear();
        viewport.draw();
    }
    else if(viewport.v_y > 0){
        viewport.v_y -= 1;
        viewport.dirty = true;
        viewport.clear();
        viewport.draw();
    }
}

viewport.down = function(){
    cursed.viewer.handling = true;
    setTimeout(()=>{cursed.viewer.handling = false;}, 100);

    if(viewport.v_y < ((viewport.v_height - viewport.height - 2))){
        viewport.v_y += 2;
        viewport.dirty = true;
        viewport.clear();
        viewport.draw();
        cursed.status_line.draw();
    }
    else if(viewport.v_y < ((viewport.v_height - viewport.height))){
        viewport.v_y += 1;
        viewport.dirty = true;
        viewport.clear();
        viewport.draw();
        cursed.status_line.draw();
    }
    
}

viewport.left = function(){
    cursed.viewer.handling = true;
    setTimeout(()=>{cursed.viewer.handling = false;}, 100);

    if(viewport.v_x > 2){
        viewport.v_x -= 2;
        viewport.dirty = true;
        viewport.clear();
        viewport.draw();
        cursed.status_line.draw();
    }
    else if(viewport.v_x > 0){
        viewport.v_x -= 1;
        viewport.dirty = true;
        viewport.clear();
        viewport.draw();
        cursed.status_line.draw();
    }
}

viewport.right = function(){
    cursed.viewer.handling = true;
    setTimeout(()=>{cursed.viewer.handling = false;}, 100);

    if(viewport.v_x < ((viewport.v_width - viewport.width - 2))){
        viewport.v_x += 2;
        viewport.dirty = true;
        viewport.clear();
        viewport.draw();
        cursed.status_line.draw();
    }
    else if(viewport.v_x < ((viewport.v_width - viewport.width))){
        viewport.v_x += 1;
        viewport.dirty = true;
        viewport.clear();
        viewport.draw();
        cursed.status_line.draw();
    }
}


viewport.updateBounds = function(x, y){
    viewport.v_height = y;
    viewport.v_width = x;
}

viewport.updateFeatures = function(features){
    viewport.features = features;
    //viewport.dirty = true;
    //viewport.clear();
    //viewport.draw();
}

viewport.updateFow = function(fow){
    viewport.fow = fow;
    //viewport.dirty = true;
    //viewport.clear();
    //viewport.draw();
}

viewport.updateUnits = function(units){
    viewport.units = units;
    //viewport.dirty = true;
    //viewport.clear();
    //viewport.draw();
};

viewport.getCursorFocus = function(){
    var unit = viewport.getCurrentUnit();
    var feature = null;



    if (unit === null){
        var i = viewport.features.length;
        // Draw features
        while(i--){
            var f = viewport.features[i];
            if(f.x == viewport.cursor_x + viewport.v_x &&
               f.y == viewport.cursor_y + viewport.v_y){
                feature = cursed.features.get(f.type);
                break;
            }
        }
    }

    if( !(  cursed.state.role === "pc" && viewport.fow[viewport.cursor_y + viewport.v_y] !== undefined && viewport.fow[viewport.cursor_y + viewport.v_y][viewport.cursor_x + viewport.v_x] )){
        
        if(unit !== null){
            var text = unit.name + " " + unit.current_health + "/" + unit.max_health;
            var bar = "";
            
            if(unit.max_health !== 0){
                var font_width = cursed.constants.font_size + cursed.constants.font_width_offset + cursed.constants.font_spacing_offset;
                var width = Math.floor((cursed.colon_line.width - ((16+text.length)*font_width))/(font_width)) - 4; // +16 for the side bars and the " " and [ ] characters

                var percent = unit.current_health/(unit.max_health*1.0);

                if(percent <= 1.0 && percent >= 0.0){
                    var number_of_units =  Math.ceil(width*percent);
                    bar = Array(number_of_units+1).join("=");
                    var width_padding = Array(width).join(" ");
                    bar = String(width_padding + bar).slice(-width);
                }
                else if(percent > 1.0){
                    var diff = percent - 1.0;
                    var diff_percent = diff/percent;
                    var excess = Math.ceil(width * diff_percent);

                    bar = Array(excess+1).join("=");
                    bar += "(+)|";

                    var width_padding = Array(width+1).join("=");
                    bar = String(bar + width_padding ).substring(0, width);
                }
                else if(percent < 0.0){
                    var diff = (-1 * percent)/(1.0 + (-1 * percent));
                    var excess = Math.ceil(width * diff);

                    bar = "|(-)"
                    bar += Array(excess+1).join("=");

                    var width_padding = Array(width).join(" ");
                    bar = String(width_padding + bar).slice(-width);
                }

            }

            return text + " [" + bar + "]";
        }
        else if (feature !== null){
            var text = feature.name;
            return text;
        }
        else{
            return "";
        }

    }
    else{
        return "";
    }
}

viewport.getCurrentUnit = function(){
    var i = viewport.units.length;
    while(i--){
        var u = viewport.units[i];
        if(u.x === (viewport.cursor_x + viewport.v_x) && u.y === (viewport.cursor_y + viewport.v_y)){
            return u;
        }
    }
    return null;
}

viewport.getUnit = function(id){
    for(var i=0; i<viewport.units.length; i++){
        if(viewport.units[i].id === id){
            return viewport.units[i];
        }
    }
    return null;
}
